airflow/providers/pinecone/provider.yaml

```diff
@@ -44,7 +44,7 @@ dependencies:
   # Pinecone Python SDK v3.0.0 was released at 2024-01-16 and introduce some breaking changes.
   # It's crucial to adhere to the v3.0.0 Migration Guide before the upper-bound limitation can be removed.
   # https://canyon-quilt-082.notion.site/Pinecone-Python-SDK-v3-0-0-Migration-Guide-056d3897d7634bf7be399676a4757c7b
-  - pinecone-client>=2.2.4,<3.0
+  - pinecone-client~=3.0.0
```

Comment:
This would not work as expected, e.g. maximum allowed version would be 3.1 - 0.0.1

You could check it by use [`packaging`](https://packaging.pypa.io/en/stable/index.html) library

```python
from packaging.requirements import SpecifierSet
from packaging.version import Version


specifiers = [
    SpecifierSet("~=3.0.0"),
    SpecifierSet("~=3.0"),
    SpecifierSet(">=3,<4"),
]

versions = [
    Version("2.2.4"),
    Version("3.0.0"),
    Version("3.0.1"),
    Version("3.1.0"),
    Version("4.0.0"),
]

for spec in specifiers:
    print(f" {spec} ".center(72, "="))
    for version in versions:
        print(f"Version {version} in spec? {version in spec}")
```

Reply:
Regarding the pin, I did want to limit >=3.0.0 and <3.1.0. Since, <2 wouldn't work for us and 3.1 can possibly have breaking changes. Once Pinecone releases 3.1, I thought we can test them out and then change pin to `pinecone-client~=3.0` . Do you think its a reasonable approach?

Reply:
If pinecone maintainers follow semver the minor/feature versions shouldn't introduce breaking changes. I can't find any information that pinecone have a plan to release breaking changes in 3.1.

Usual we do not use upper bound limits, only in edge cases: e.g. it is known fact that new major (rarely minor) of SDK will introduce breaking changes.

Reply:
I understand. Thanks,  I'll use `pinecone-client>=3.0.0`